A39-xds-http-filters.md

```diff
@@ -0,0 +1,152 @@
+  [`type_url`](https://github.com/cncf/udpa/blob/cc1b757b3eddccaaaf0743cbb107742bb7e3ee4f/udpa/type/v1/typed_struct.proto#L38)
+  field to determine the actual type; this allows for third-party
+  plugins to encode their configurations as a `google.protobuf.Struct`
+  proto instead of using a concrete proto message type that the client
+  may not be able to handle.
+
+When gRPC receives the `HttpConnectionManager` proto in the LDS
+response, it will validate the list of filters as follows:
+- Every filter in the list must be of a known type, as identified by the
+  `type_url` field.  Any unknown config type will cause the Listener
+  resource to be NACKed.
+- Every filter in the list must have a unique instance name.  Any
+  duplicate instance names will cause the Listener resource to be NACKed.
+
+Note that gRPC will support configuring multiple instances of the same
+filter implementation, each with its own name and config.
+
+In each of the `VirtualHost`, `Route`, and `WeightedCluster.ClusterWeight`
+protos, there is a field called `typed_per_filter_config`, which is a map
+from filter instance name to a corresponding filter-specific configuration
+proto message.  (Note that Envoy currently keys these maps by legacy
+filter names instead of by filter instance names, but this is expected
+to change, as per https://github.com/envoyproxy/envoy/issues/12274; gRPC
+will directly implement the desired semantics here.)
+
+When gRPC receives the `RouteConfiguration` proto, either in the LDS or
+RDS response, it will validate the contents of each
+`typed_per_filter_config` map as follows:
+- Every config in the list must be of a known type, as identified by the
+  map value proto message type.  Any unknown config type will cause the
+  resource to be NACKed.
+- Note that gRPC will *not* fail validation if the map key specifies a
+  filter instance name that does not exist in the `HttpConnectionManager`
+  filter list.  This is because during an update, the xDS client code
+  cannot know which `HttpConnectionManager` config is currently being
+  used.
+
+Unlike Envoy, which provides an API that filters must call to access their
+per-{VirtualHost,Route,ClusterWeight} config overrides, gRPC will
+require filters to construct a merged configuration, applying any
+necessary overrides, when the xDS config is applied.  The filters will
+then use this merged configuration at run-time.  However, this approach
+is an implementation detail that could change in the future.
+
+### Initially Supported Filter: Router
+
+Initially, gRPC will support only one filter, the [router
+filter](https://www.envoyproxy.io/docs/envoy/latest/configuration/http/http_filters/router_filter),
+which is used for the proto message type
+[`extensions.filters.http.router.v3.Router`](https://github.com/envoyproxy/envoy/blob/18db4c90e3295fb2c39bfc7b2ce641cfd6c3fbed/api/envoy/extensions/filters/http/router/v3/router.proto#L23).
```

Comment:
Do we need to support the router config being encoded inside a `udpa.type.v1.TypedStruct`, or will it always be directly in the `typed_config` field?

Reply:
> both Envoy and in my code for C-core in grpc/grpc#25310 ... just uses the resulting type to look up the filter implementation that should parse the proto

So what's the end result for Envoy/C right now when this happens?

You validate LDS successfully, but then the actual filter implementation fails because it finds the wrong message type for the config?  So...RPCs fail?  Or since we don't actually support any of the configuration for the `Router` type anyway, the whole config can be safely ignored, and it "works" (for now)?  Or I could also see a possible implementation where we end up processing any filters after this one, and if we find another `Router` entry, everything "works" but this oddly-encoded filter is just ignored.

Most importantly: what do we _want_ to do?  I can implement it any of these ways with roughly the same effort - I just need to know what to do.

Reply:
Currently, the router filter impl does not expect its config as a `TypedStruct`, so the config will be rejected.  But it won't be rejected by the general HCM parsing code; it will be rejected by the router filter.  Specifically, here's what the HCM parsing code will do:

1. Look at the type of the filter config.  The type is `TypedStruct`, so it will look inside of the `TypedStruct` for the real name, where it will see `envoy.extensions.filters.http.router.v3.Router`.
2. Ask the registry what filter to use for config type `envoy.extensions.filters.http.router.v3.Router`.  The registry will return the router filter.
3. Ask the router filter to parse the `TypedStruct`.  The router filter will return an error, because it doesn't know how to parse the `TypedStruct`.
4. Because of that error, the HCM parsing code will reject the config.

Reply:
It's not specified in the proposal when the configs should be parsed by the filters.  I was assuming we would wait until the entire config was known (the merge of the 4 sources) before passing it to the filter.  Is that wrong?  Can you document exactly how and when parsing should occur?

Reply:
I think the proposal already states this explicitly.  It says:

> When gRPC receives the `HttpConnectionManager` proto in the LDS
> response, it will validate the list of filters as follows:
> - Every filter in the list must be of a known type, as identified by the
>   `type_url` field.  Any unknown config type will cause the Listener
>   resource to be NACKed.
> - Every filter in the list must have a unique instance name.  Any
>   duplicate instance names will cause the Listener resource to be NACKed.

And then later:

> When gRPC receives the `RouteConfiguration` proto, either in the LDS or
> RDS response, it will validate the contents of each
> `typed_per_filter_config` map as follows:
> - Every config in the list must be of a known type, as identified by the
>   map value proto message type.  Any unknown config type will cause the
>   resource to be NACKed.
> - Note that gRPC will *not* fail validation if the map key specifies a
>   filter instance name that does not exist in the `HttpConnectionManager`
>   filter list.  This is because during an update, the xDS client code
>   cannot know which `HttpConnectionManager` config is currently being
>   used.

Reply:
As we discussed offline, I've clarified that the filter config itself will be validated, and I've added a section about the override config merging precedence rules.

Reply:
What happens when a single filter name appears with multiple, incompatible types?

I.e.:
```
LDS:
HttpConnectionManager.http_filters: {
  name: "FOO_FILTER"
  typed_config: extensions.filters.http.fault.v3.HTTPFault {
    // fault injection config
  }
}

RDS:
VirtualHost.typed_per_filter_config["FOO_FILTER"]: extensions.filters.http.health_check.v3.HealthCheck {
  // health check config
}
```

IIUC, with this proposal, assuming gRPC has implemented fault injection and health checking:

- LDS will validate, as the fault injection filter is able to parse the HTTPFault config.
- RDS will validate, as the health check filter is able to parse the HealthCheck config.
- When an RPC happens which uses that VirtualHost, we will call the fault injection HTTP filter, passing it the FI config from HCM and the HC config from the VH.  This is not valid - so RPCs will fail with an appropriate error (something like "unexpected config type") when this happens?

Is this what happens in Envoy?

Or should it be caught earlier and nack the whole config (LDS, I guess)?

Reply:
Yeah, unfortunately, I don't think we'll be able to NACK in that case.

Unfortunately, as we've discussed in other contexts, the xDS ACK/NACK is really just about the validation of an individual resource in isolation, not about whether it actually works in the context of the rest of the client's configuration.  xDS does not currently have any mechanism for the client to indicate its overall config health, so there's really no way to handle this kind of case.  One obvious example of where this happens in Envoy is if it receives two Listeners for the same port.  Another case that affects both Envoy and gRPC would be if a RouteConfiguration refers to a Cluster that does not actually exist.  In these kinds of cases, we'll ACK but not actually work, and the case you describe here is just another example of that general problem.  See envoyproxy/envoy#11396 for discussion of one possible solution.

I think the case you cite here is another example of this problem.  In terms of layering, the validation code (inside of the `XdsClient`) doesn't actually know which LDS resource a given RDS resource will be used with -- and even if there weren't layering issues to consider, if both the LDS and RDS resources are being updated at around the same time, it would be really hard to figure out *which* LDS resource the RDS resource is supposed to match up with.  This is basically why validation only looks at each individual resource in isolation.

I think Envoy will do the same thing once they implement the changes described in envoyproxy/envoy#12274, and I've already added a note about this to that issue.

Reply:
That all sounds reasonable. So what should happen, then?  RPCs that end up invoking this configuration combination just fail?

Reply:
I think it's ultimately up to the filter how to handle it.  In C-core, we'll just pass the two configs to the filter's `GenerateServiceConfig()` method, which will notice that the override config is of the wrong type.

My off-the-cuff guess is that the right thing to do here is to (a) log an error indicating the problem and (b) ignore the override config.  But individual filters might choose to do something different for their own reasons.

Reply:
So for the specific design for the router filter in this gRFC, we should log an error but ignore the misconfiguration?

Reply:
In the specific case of the router filter, it doesn't accept per-route overrides, so its `GenerateFilterConfigOverride()` method should unconditionally return an error, regardless of what its input was.